docs/root/intro/life_of_a_request.rst

```diff
@@ -0,0 +1,631 @@
+
+We use the terms :ref:`listener subsystem <arch_overview_listeners>` and :ref:`cluster subsystem
+<arch_overview_cluster_manager>` above to refer to the group of modules and instance classes that
+are created by the top level `ListenerManager` and `ClusterManager` classes. There are many
+components that we discuss below that are instantiated before and during the course of a request by
+these management systems, for example listeners, filter chains, codecs, connection pools and load
+balancing data structures.
+
+Envoy has an `event-based thread model
+<https://blog.envoyproxy.io/envoy-threading-model-a8d44b922310>`_. A main thread is responsible for
+the server lifecycle, configuration processing, stats, etc. and some number of :ref:`worker threads
+<arch_overview_threading>` process requests. All threads operate around an event loop (`libevent
+<https://libevent.org/>`_) and any given downstream TCP connection (including all the multiplexed
+streams on it) will be handled by exactly one worker thread for its lifetime. Each worker thread
+maintains its own pool of TCP connections to upstream endpoints. :ref:`UDP
+<arch_overview_listeners_udp>` handling makes use of SO_REUSEPORT to have the kernel consistency
+hash the source/destination IP:port tuples to the same worker thread. UDP filter state is global to
+a worker thread, with the filter responsible for providing session semantics as needed. This is in
+contrast to the connection oriented TCP filters we discuss below.
+
+Worker threads rarely share state and operate in a trivially parallel fashion. This threading model
+enables scaling to very high core count CPUs.
+
+Request flow
+------------
+
+Overview
+^^^^^^^^
+
+A brief outline of the life cycle of a request and response using the example configuration above:
+
+1. A TCP connection from downstream is accepted by an Envoy :ref:`listener
+   <arch_overview_listeners>` running on a :ref:`worker thread <arch_overview_threading>`.
+2. The :ref:`listener filter <arch_overview_listener_filters>` chain is created and runs. It can
+   provide SNI and other pre-TLS info. Once completed, the listener will match a network filter
+   chain. Each listener may have multiple filter chains which match on some combination of
+   destination IP CIDR range, SNI, ALPN, source ports, etc. A transport socket, in our case the TLS
+   transport socket, is associated with this filter chain.
+3. On network reads, the :ref:`TLS <arch_overview_ssl>` transport socket decrypts the data read from
+   the TCP connection to plain text for further processing.
+4. The :ref:`network filter <arch_overview_network_filters>` chain is created and runs. The most
+   important filter for HTTP is the HTTP connection manager, which is the last network filter in the
+   chain.
+5. The HTTP/2 codec in :ref:`HTTP connection manager <arch_overview_http_conn_man>` demultiplexes
+   the plain text TCP connection to a number of independent HTTP streams. Each HTTP stream handles a
```

Comment:
"plain text TCP connection" is confusing because all TCP connections in the example are TLS. Maybe "plain text data stream", maybe add "from the decrypted TLS connection".

Also, for people not familiar with encryption terminology, it may not be clear that "plain text" doesn't mean text as in HTTP/1 is a text-based protocol and HTTP/2 is a binary protocol. So maybe instead of "plain text data stream" use "decrypted data stream".